### PR TITLE
[ADD] website_sale_qty: Website sale price tiers

### PR DESCRIPTION
**Pull request by obulkin**
- Add the calculated field price_quantity_tiers to product.template. It stores
  a dictionary with price tiers (quantity ranges with different unit pricing) by
  product.pricelist
- Update the website_sale.product template to show radio buttons corresponding
  to the price tiers for each product.template/product.pricelist combo
- Add JS that links the new price tier radio buttons to the existing quantity
  selection elements

### CODE REVIEW

**Comment by lasley** [comment]
cc @laslabs

File: website_sale_qty/__openerp__.py
Comment (lasley) [inline website_sale_qty/__openerp__.py]: remove empty keys

File: website_sale_qty/__openerp__.py
Comment (lasley) [inline website_sale_qty/__openerp__.py]: Please switch all of this to LGPL-3. Get with me in chat when you have some time so we can discuss licensing

File: website_sale_qty/__openerp__.py
Comment (lasley) [inline website_sale_qty/__openerp__.py]: base is inferred, remove

File: website_sale_qty/models/product_template.py
Comment (lasley) [inline website_sale_qty/models/product_template.py]: `fields.Date.today()`

File: website_sale_qty/models/product_template.py
Comment (lasley) [inline website_sale_qty/models/product_template.py]: `fields.Date.today()`

File: website_sale_qty/models/product_template.py
Comment (lasley) [inline website_sale_qty/models/product_template.py:30]: You can chain the ors to make this look better. `'|', '|',` Then remove the Ln31

File: website_sale_qty/models/product_template.py
Comment (lasley) [inline website_sale_qty/models/product_template.py]:
```
in_quantities.add(
    1 if price.min_quantity < 1 else price.min_quantity
)
```

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [inline website_sale_qty/static/src/js/website_sale_qty.js]: Use camelCase variables in JS where we would normally snake_case in Python/Ruby.

JS keeps with UpperCamelCase in the same spots we'd use UpperCamelCase in Python/Ruby.

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [inline website_sale_qty/static/src/js/website_sale_qty.js:9]: `var _ = require('_');` - see my comment regarding underscore later for why

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [inline website_sale_qty/static/src/js/website_sale_qty.js]: Use underscore for a loop without a frivolous var  - `_.each` is reversed (val, key) vs jQuery (key, val)

```
_.each(radios, function(element) {
```

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [inline website_sale_qty/static/src/js/website_sale_qty.js:8]: `var $ = require('$')` - even though it works, you haven't guaranteed that jQuery exists in the pipeline before using in L19&20

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [inline website_sale_qty/static/src/js/website_sale_qty.js]: `event.currentTarget`

**Review by lasley — CHANGES_REQUESTED** [review]
Thanks @obulkin

**Comment by lasley** [comment]
Seems we have a missing trigger
- Navigate to Ipad Mini
- Quantity to 5
- Price stays at $320
- Hit `-` button on qty field (to drop to qty 4)
- Hit `+` button on qty field (to go back to 5)
- Price is now $280 with $320 crossed out.

File: website_sale_qty/models/product_template.py
Comment (obulkin) [reply website_sale_qty/models/product_template.py:30]: I think that would be logically different. The current set of domains works out to:

```
pricelist_criterion and product_tmpl_criterion 
and (date_start_criterion_1 or date_start_criterion_2) 
and (date_end_criterion_1 or date_end_criterion_2)
```

But with the change, it would be:

```
pricelist_criterion and product_tmpl_criterion 
and ((date_start_criterion_1 or date_start_criterion_2) or date_end_criterion_1)
and date_end_criterion_2
```

My testing seems to confirm this.

File: website_sale_qty/models/product_template.py
Comment (obulkin) [reply website_sale_qty/models/product_template.py]: Is there any upside to calling this method instead?

Comment (obulkin) [reply website_sale_qty/models/product_template.py]: I suppose it'll eliminate the need for an extra import. Is that what you had in mind?

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (obulkin) [reply website_sale_qty/static/src/js/website_sale_qty.js:8]: So the presence of jQuery and Underscore globals isn't guaranteed as part of v9? I see a number of other v9 modules that aren't going through the `require` process for jQuery access.

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (obulkin) [reply website_sale_qty/static/src/js/website_sale_qty.js]: Good to know. If the goal is just to eliminate the unnecessary variable then I can also do the following, which seems like the best solution because it doesn't involve bringing in an extra library:

```
radios.each(function(index) {
    var $element = $(radios[index]);
    radio_map[$element.val()] = $element;
});
```

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (obulkin) [reply website_sale_qty/static/src/js/website_sale_qty.js]: Good catch. At least in Chrome, `event.target` works just fine here because of the way that a click on a `<label>` element produces a click on the associated `<input>`, but `event.currentTarget` would be more universal.

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [reply website_sale_qty/static/src/js/website_sale_qty.js]: `_` is already included in Odoo, so you should use it. You will need to `var _ = require('_');` though

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [reply website_sale_qty/static/src/js/website_sale_qty.js:8]: You cannot guarantee the presence of something you did not require. The other modules are wrong if they did not require things before using

File: website_sale_qty/models/product_template.py
Comment (lasley) [reply website_sale_qty/models/product_template.py]: You're putting an object in a domain. That's invalid. It should be a string in the proper format, which is what fields does.

File: website_sale_qty/models/product_template.py
Comment (lasley) [reply website_sale_qty/models/product_template.py:30]: Works for me

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [reply website_sale_qty/static/src/js/website_sale_qty.js]: Chrome will also show issues with target vs curentTarget, but the issues are few and far between. Just a best practice thing really

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (obulkin) [reply website_sale_qty/static/src/js/website_sale_qty.js:8]: Fair enough. I'll do this moving forward.

File: website_sale_qty/models/product_template.py
Comment (obulkin) [reply website_sale_qty/models/product_template.py]: Ah. Got it. The relevant tests seem to pass either way, but I'm happy to change it.

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (obulkin) [reply website_sale_qty/static/src/js/website_sale_qty.js]: I understand that, but there must be some overhead associated with each call to `require`, and my suggested correction avoids that while also eliminating the unused variable. Is there a problem with it?

**Comment by obulkin** [comment]
@lasley --- Good catch on the missing trigger. I've updated the code to fire a `change` event after the quantity input is updated via `val()`, which takes care of the issue.

File: website_sale_qty/static/src/js/website_sale_qty.js
Comment (lasley) [reply website_sale_qty/static/src/js/website_sale_qty.js]: The overhead is negligible. Using the underscore creates cleaner code. Please do it.

**Comment by lasley** [comment]
Excellent thanks @obulkin 

I don't see new commits. Did you forget to push the code, perhaps?

**Comment by obulkin** [comment]
@lasley --- I wanted to push everything up at once and was waiting for clarity around the inline comment about Underscore. The changes should be up soon.

**Comment by lasley** [comment]
@obulkin - Please don't squash your commits until after reviews are over. This entire PR needs to be re-reviewed now, instead of just the diff.

**Review by lasley — APPROVED** [review]
code + test 👍

**Comment by obulkin** [comment]
@lasley --- My bad. I'll hold off on squashing from now on.

**Comment by lasley** [comment]
@obulkin - Request from customer, please make the price readout on the main product page say "Per `product.uom_id.name`":

image

File: website_sale_qty/__openerp__.py
Comment (oscarolar) [inline website_sale_qty/__openerp__.py:10]: Hello, can you change the website key according to this guideline https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L9

File: website_sale_qty/__openerp__.py
Comment (oscarolar) [inline website_sale_qty/__openerp__.py]: Hello, can you change the authors key according to this guideline:
https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L10

File: website_sale_qty/__openerp__.py
Comment (pedrobaeza) [reply website_sale_qty/__openerp__.py:10]: Website key is allowed to be the company one

File: website_sale_qty/__openerp__.py
Comment (pedrobaeza) [reply website_sale_qty/__openerp__.py]: This comment is correct. It should include "Odoo Community Association (OCA)" for indexing purposes.

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (oscarolar) [inline website_sale_qty/static/tests/js/website_sale_qty.tour.js:98]: I don't think including `if` statements in tests is a good practice, can't this be replaced by?:

``` js
waitFor: '.js-pqt-input[value="5"]:checked', 
```

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (oscarolar) [inline website_sale_qty/static/tests/js/website_sale_qty.tour.js:84]: I don't think including `if` statements in tests is a good practice, can't this be replaced by?:

``` js
waitFor: '.js-pqt-input:not(:checked)', 
```

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (oscarolar) [inline website_sale_qty/static/tests/js/website_sale_qty.tour.js:70]: I don't think including `if` statements in tests is a good practice, can't this be replaced by?:

``` js
waitFor: '.js-pqt-input[value="5"]:checked', 
```

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (oscarolar) [inline website_sale_qty/static/tests/js/website_sale_qty.tour.js:53]: I don't think including `if` statements in tests is a good practice, can't this be replaced by?:

``` js
waitFor: '.js-pqt-input:not(:checked)', 
```

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (oscarolar) [inline website_sale_qty/static/tests/js/website_sale_qty.tour.js:22]: Same for this case

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (oscarolar) [inline website_sale_qty/static/tests/js/website_sale_qty.tour.js:36]: same for this case.

**Review by oscarolar — COMMENTED** [review]
I think it this all if inside the test are not necessary, you can do all that with `waitFor` and `waitNot` keys.

File: website_sale_qty/__openerp__.py
Comment (lasley) [reply website_sale_qty/__openerp__.py]: @obulkin - please heed this advice for all of our submissions to OCA

**Comment by lasley** [comment]
It should not show the radio button when there is only one price tier

image

**Comment by lasley** [comment]
Error 500 when no pricelist tier available:

```
2016-10-17 23:59:01,684 32339 ERROR test_db openerp.addons.website.models.ir_http: 500 Internal Server Error:
...
  File "/opt/odoo/community/oca-e-commerce/website_sale_qty/models/product_template.py", line 44, in _compute_price_quantity_tiers
    record.product_variant_ids[0].id, min_qty
  File "/media/enc-data-02/odoo/openerp/models.py", line 5761, in __getitem__
    return self._browse(self.env, (self._ids[key],))
QWebException: "tuple index out of range" while evaluating
'product.price_quantity_tiers[pricelist.id]'
```

File: website_sale_qty/static/tests/js/website_sale_qty.tour.js
Comment (obulkin) [reply website_sale_qty/static/tests/js/website_sale_qty.tour.js:22]: Actually, the `waitFor`/`waitNot` approach doesn't seem to work properly here. For example, the tour will still finish successfully even if this step includes a line like `waitFor: '.js-pqt-input[value="1"]:not(:checked)'`. Something similar happens with the last step in the tour, where `waitNot: '.js-pqt-input[value="5"]:checked'` doesn't generate an error. And I don't think there's any way to check the value **property** of an `<input>` as part of a jQuery selector, which would be necessary to replace the `if` statement in the step below.

In light of all these issues, I decided to just go with the `onload` callback throughout for consistency.

**Comment by obulkin** [comment]
Thanks for the feedback, everyone. I just pushed up a new commit that should address all of this with the exception of the tour feedback left by @oscarolar. I've commented on that inline.

**Review by lasley — APPROVED** [review]
Thanks @obulkin

**Comment by lasley** [comment]
Any PSC with thoughts on this? I'd like to see this merged soon

**Comment by pedrobaeza** [comment]
What about the comments from @oscarolar ?

**Comment by obulkin** [comment]
@pedrobaeza - I addressed those [in line](https://github.com/OCA/e-commerce/pull/131#discussion_r83781815).

File: website_sale_qty/README.rst
Comment (pedrobaeza) [inline website_sale_qty/README.rst]: Don't add these "empty" sections.

**Review by pedrobaeza — APPROVED** [review]
Little comment, but overall good (code review)

File: website_sale_qty/README.rst
Comment (obulkin) [reply website_sale_qty/README.rst]: Sure thing. I'll drop them. I figured I needed to put something there to align with the OCA template.

**Comment by lasley** [comment]
Looks like we're good to merge?

File: website_sale_qty/README.rst
Comment (elicoidal) [inline website_sale_qty/README.rst]: I miss the runbot section

**Review by elicoidal — APPROVED** [review]
minor non-blocking in the README

File: website_sale_qty/README.rst
Comment (obulkin) [reply website_sale_qty/README.rst]: Good catch. I've added a Runbot link.

**Comment by lasley** [comment]
Good to merge?

**Pull request merged**
